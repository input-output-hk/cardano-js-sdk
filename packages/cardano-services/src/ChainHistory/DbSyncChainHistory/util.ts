import { Cardano, PaginationArgs } from '@cardano-sdk/core';
import {
  CertificateModel,
  DelegationCertModel,
  MirCertModel,
  PoolRegisterCertModel,
  PoolRetireCertModel,
  StakeCertModel,
  WithCertType
} from './types';

export const isPoolRetireCertModel = (
  value: WithCertType<CertificateModel>
): value is WithCertType<PoolRetireCertModel> => value.type === 'retire';

export const isPoolRegisterCertModel = (
  value: WithCertType<CertificateModel>
): value is WithCertType<PoolRegisterCertModel> => value.type === 'register';

export const isMirCertModel = (value: WithCertType<CertificateModel>): value is WithCertType<MirCertModel> =>
  value.type === 'mir';

export const isStakeCertModel = (value: WithCertType<CertificateModel>): value is WithCertType<StakeCertModel> =>
  value.type === 'stake';

export const isDelegationCertModel = (
  value: WithCertType<CertificateModel>
): value is WithCertType<DelegationCertModel> => value.type === 'delegation';

export const applyPagination = (ids: Cardano.TransactionId[], { startAt, limit }: PaginationArgs) =>
  ids.slice(startAt).slice(0, limit);
