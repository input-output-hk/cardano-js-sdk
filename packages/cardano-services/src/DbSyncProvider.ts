import { HealthCheckResponse, Provider } from '@cardano-sdk/core';
import { Pool } from 'pg';

const HEALTH_CHECK_QUERY = 'SELECT 1';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyArgs = any[];

export const DbSyncProvider = <
  T extends (abstract new (...args: AnyArgs) => {}) | (new (...args: AnyArgs) => {}) = { new (): {} }
>(
  BaseClass?: T
) => {
  abstract class Mixin extends (BaseClass || Object) implements Provider {
    public db: Pool;

    constructor(...args: AnyArgs) {
      const [db, ...baseArgs] = [...args];

      super(...baseArgs);

      this.db = db;
    }

    public async healthCheck(): Promise<HealthCheckResponse> {
      const result = await this.db.query(HEALTH_CHECK_QUERY);
      return { ok: !!result.rowCount };
    }
  }

  type BaseArgs = T extends abstract new (...baseArgs: infer A) => {}
    ? A
    : T extends new (...baseArgs: infer A) => {}
    ? A
    : never;
  type BaseInstance = T extends abstract new (...baseArgs: AnyArgs) => infer I
    ? I
    : T extends new (...baseArgs: AnyArgs) => infer I
    ? I
    : never;
  type ReturnedType = BaseInstance & { db: Pool; healthCheck: () => Promise<HealthCheckResponse> };

  return Mixin as unknown as (T extends new (...baseArgs: AnyArgs) => {}
    ? new (db: Pool, ...args: BaseArgs) => ReturnedType
    : abstract new (db: Pool, ...args: BaseArgs) => ReturnedType) & {
    prototype: { healthCheck: () => Promise<HealthCheckResponse> };
  };
};

export type DbSyncProvider<T> = T extends new (...args: AnyArgs) => infer O ? O : never;
