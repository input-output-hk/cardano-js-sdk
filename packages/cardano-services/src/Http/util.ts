import { ProviderError, ProviderFailure } from '@cardano-sdk/core';

export const ORIGIN = 'Origin';
export const CONTENT_TYPE = 'Content-Type';
export const APPLICATION_JSON = 'application/json';

export const getListen = (url: URL) => ({ host: url.hostname, port: Number.parseInt(url.port) });

type StaticOrigin = boolean | string | RegExp | (boolean | string | RegExp)[];

export const corsOptions = (allowedOrigins: Set<string>) => ({
  origin(requestOrigin: string | undefined, callback: (err: Error | null, options?: StaticOrigin) => void) {
    if (requestOrigin && allowedOrigins.has(requestOrigin)) {
      callback(null, requestOrigin);
    } else {
      callback(new ProviderError(ProviderFailure.Forbidden, null, `Origin ${requestOrigin} not allowed by CORS`));
    }
  }
});
