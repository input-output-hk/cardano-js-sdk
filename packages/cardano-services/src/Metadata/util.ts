import { Cardano } from '@cardano-sdk/core';
import { TxMetadataByHashes } from './DbSyncMetadataService';
import { TxMetadataModel } from './types';
import { mapTxMetadata } from './mappers';

export const mapTxMetadataByHashes = (listOfMetadata: TxMetadataModel[]): TxMetadataByHashes => {
  const metadataMap: Map<Cardano.TransactionId, TxMetadataModel[]> = new Map();

  for (const metadata of listOfMetadata) {
    const txId = metadata.tx_id.toString('hex') as Cardano.TransactionId;
    const currentMetadata: TxMetadataModel[] = metadataMap.get(txId) ?? [];
    metadataMap.set(txId, [...currentMetadata, metadata]);
  }

  return new Map([...metadataMap].map(([id, metadata]) => [id, mapTxMetadata(metadata)]));
};
