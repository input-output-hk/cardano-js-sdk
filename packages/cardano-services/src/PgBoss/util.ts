import { PgBossQueue, workerQueues } from './types';

export const isValidQueue = (queue: string): queue is PgBossQueue => workerQueues.includes(queue as PgBossQueue);

export const isErrorWithConstraint = (error: unknown): error is Error & { constraint: unknown } =>
  error instanceof Error && 'constraint' in error;
