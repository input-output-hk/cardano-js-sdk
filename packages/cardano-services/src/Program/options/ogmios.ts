import { Command, Option } from 'commander';
import { Ogmios } from '@cardano-sdk/ogmios';
import { URL } from 'url';

const OGMIOS_URL_DEFAULT = (() => {
  const connection = Ogmios.createConnectionObject();
  return connection.address.webSocket;
})();

export enum OgmiosOptionDescriptions {
  SrvServiceName = 'Ogmios SRV service name',
  Url = 'Ogmios URL'
}

export interface OgmiosProgramOptions {
  ogmiosUrl?: URL;
  ogmiosSrvServiceName?: string;
}

export const withOgmiosOptions = (command: Command) =>
  command
    .addOption(
      new Option('--ogmios-srv-service-name <ogmiosSrvServiceName>', OgmiosOptionDescriptions.SrvServiceName).env(
        'OGMIOS_SRV_SERVICE_NAME'
      )
    )
    .addOption(
      new Option('--ogmios-url <ogmiosUrl>', OgmiosOptionDescriptions.Url)
        .env('OGMIOS_URL')
        .default(new URL(OGMIOS_URL_DEFAULT))
        .conflicts('ogmiosSrvServiceName')
        .argParser((url) => new URL(url))
    );
