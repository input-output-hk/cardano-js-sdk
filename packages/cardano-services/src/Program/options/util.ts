import { Option } from 'commander';
import type { Command } from 'commander';

export const addOptions = (command: Command, options: Option[]) => {
  for (const option of options) command.addOption(option);

  return command;
};

export const newOption = <T>(
  flags: string,
  description: string,
  envName: string,
  argParser?: (value: string, previous: T) => T,
  defaultValue?: unknown
) => {
  const option = new Option(flags, description).env(envName);

  if (argParser !== undefined) option.argParser(argParser);
  if (defaultValue !== undefined) option.default(defaultValue);

  return option;
};
