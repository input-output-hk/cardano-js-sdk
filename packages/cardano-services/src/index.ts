export * from './Asset';
export * from './Http';
export * from './InMemoryCache';
export * from './Program';
export * from './RunnableModule';
export * from './StakePool';
export * from './TxSubmit';
export * from './Utxo';
export * from './ChainHistory';
export * from './Rewards';
export * as errors from './errors';
export * as util from './util';
