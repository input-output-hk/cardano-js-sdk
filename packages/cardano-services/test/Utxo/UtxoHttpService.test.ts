/* eslint-disable max-len */
import { BlockNoModel, findLastBlockNo } from '../../src/util/DbSyncProvider';
import { Cardano, ProviderError, ProviderFailure, UtxoProvider } from '@cardano-sdk/core';
import { CreateHttpProviderConfig, utxoHttpProvider } from '@cardano-sdk/cardano-services-client';
import { DbSyncUtxoProvider, HttpServer, HttpServerConfig, UtxoHttpService } from '../../src';
import { INFO, createLogger } from 'bunyan';
import { OgmiosCardanoNode } from '@cardano-sdk/ogmios';
import { Pool } from 'pg';
import { getPort } from 'get-port-please';
import { healthCheckResponseMock, mockCardanoNode } from '../../../core/test/CardanoNode/mocks';
import { logger } from '@cardano-sdk/util-dev';
import axios from 'axios';

const APPLICATION_JSON = 'application/json';
const APPLICATION_CBOR = 'application/cbor';
const UNSUPPORTED_MEDIA_STRING = 'Request failed with status code 415';
const BAD_REQUEST_STRING = 'Request failed with status code 400';
const toCardanoAddresses = (addresses: string[]) => addresses.map((a) => Cardano.Address(a));
describe('UtxoHttpService', () => {
  let dbConnection: Pool;
  let httpServer: HttpServer;
  let utxoProvider: DbSyncUtxoProvider;
  let service: UtxoHttpService;
  let port: number;
  let baseUrl: string;
  let clientConfig: CreateHttpProviderConfig<UtxoProvider>;
  let config: HttpServerConfig;
  let cardanoNode: OgmiosCardanoNode;
  let provider: UtxoProvider;
  let lastBlockNoInDb: Cardano.BlockNo;

  beforeAll(async () => {
    port = await getPort();
    baseUrl = `http://localhost:${port}/utxo`;
    clientConfig = { baseUrl, logger: createLogger({ level: INFO, name: 'unit tests' }) };
    config = { listen: { port } };
    dbConnection = new Pool({ connectionString: process.env.POSTGRES_CONNECTION_STRING });
  });

  describe('unhealthy UtxoProvider', () => {
    beforeEach(async () => {
      utxoProvider = {
        healthCheck: jest.fn(() => Promise.resolve({ ok: false })),
        utxoByAddresses: jest.fn()
      } as unknown as DbSyncUtxoProvider;
    });
    it('should not throw during service create if the UtxoProvider is unhealthy', () => {
      expect(() => new UtxoHttpService({ logger, utxoProvider })).not.toThrow(
        new ProviderError(ProviderFailure.Unhealthy)
      );
    });

    it('throws during service initialization if the UtxoProvider is unhealthy', async () => {
      service = new UtxoHttpService({ logger, utxoProvider });
      httpServer = new HttpServer(config, { logger, runnableDependencies: [], services: [service] });
      await expect(httpServer.initialize()).rejects.toThrow(new ProviderError(ProviderFailure.Unhealthy));
    });
  });

  describe('healthy state', () => {
    beforeAll(async () => {
      lastBlockNoInDb = (await dbConnection.query<BlockNoModel>(findLastBlockNo)).rows[0].block_no;
      cardanoNode = mockCardanoNode(
        healthCheckResponseMock({ blockNo: lastBlockNoInDb })
      ) as unknown as OgmiosCardanoNode;
      utxoProvider = new DbSyncUtxoProvider({ cardanoNode, db: dbConnection, logger });
      service = new UtxoHttpService({ logger, utxoProvider });
      provider = utxoHttpProvider(clientConfig);
      httpServer = new HttpServer(config, { logger, runnableDependencies: [cardanoNode], services: [service] });
      await httpServer.initialize();
      await httpServer.start();
    });
    afterAll(async () => {
      await dbConnection.end();
      await httpServer.shutdown();
    });

    describe('/health', () => {
      it('forwards the utxoProvider health response with HTTP request', async () => {
        const res = await axios.post(`${baseUrl}/health`, undefined, {
          headers: { 'Content-Type': APPLICATION_JSON }
        });
        expect(res.status).toBe(200);
        expect(res.data).toEqual(healthCheckResponseMock({ blockNo: lastBlockNoInDb }));
      });

      it('forwards the utxoProvider health response with provider client', async () => {
        const response = await provider.healthCheck();
        expect(response).toEqual(healthCheckResponseMock({ blockNo: lastBlockNoInDb }));
      });
    });

    describe('/utxo-by-addresses', () => {
      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await axios.post(
            `${baseUrl}/utxo-by-addresses`,
            { args: [[]] },
            { headers: { 'Content-Type': APPLICATION_CBOR } }
          );
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
          expect(error.response.status).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });

      it('returns 400 coded respons if the request is bad formed', async () => {
        try {
          await axios.post(`${baseUrl}/utxo-by-addresses`, { addresses: ['asd'] });
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
          expect(error.response.status).toBe(400);
          expect(error.message).toBe(BAD_REQUEST_STRING);
        }
      });

      it('valid request should pass OpenApi schema validations', async () => {
        const addresses = ['asd'];
        const res = await axios.post(`${baseUrl}/utxo-by-addresses`, { addresses });
        expect(res.status).toEqual(200);
      });

      it('return UTxOs for a single address', async () => {
        const res = await provider.utxoByAddresses({
          addresses: [
            Cardano.Address(
              'addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7'
            )
          ]
        });
        expect(res).toMatchSnapshot();
      });

      it('return UTxOs for multiple addresses', async () => {
        const addresses = [
          'addr_test1qp620qa3rqzd5fxj3hy4dughv7xx2dt9gu9de70jf8hagdcvmqt35f2psxv7ajj5jnh4ajlc752rert8f9msffxdl45qyjefw8',
          'addr_test1qryz24mkq35j8s67fdrm44pe8na7n3tqkmyzy3sgnjq3d7szlx56h6fkjl8y3p73zpyce04eku9w943rcr6rgznp8cwq2axy9q',
          'addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7'
        ];
        const res = await provider.utxoByAddresses({ addresses: toCardanoAddresses(addresses) });
        expect(res).toMatchSnapshot();
      });

      it('returns UTxOs containing multiple assets', async () => {
        const addresses = [
          'addr_test1qp620qa3rqzd5fxj3hy4dughv7xx2dt9gu9de70jf8hagdcvmqt35f2psxv7ajj5jnh4ajlc752rert8f9msffxdl45qyjefw8',
          'addr_test1qryz24mkq35j8s67fdrm44pe8na7n3tqkmyzy3sgnjq3d7szlx56h6fkjl8y3p73zpyce04eku9w943rcr6rgznp8cwq2axy9q'
        ];
        const res = await provider.utxoByAddresses({ addresses: toCardanoAddresses(addresses) });
        expect(res).toMatchSnapshot();
      });

      it('returns UTxOs containing multiple assets and one of the assets has no name', async () => {
        const addressAssociatedWithUTxOWithNoAssetName =
          'addr_test1qp620qa3rqzd5fxj3hy4dughv7xx2dt9gu9de70jf8hagdcvmqt35f2psxv7ajj5jnh4ajlc752rert8f9msffxdl45qyjefw8';
        const assetWithNoNameId = '126b8676446c84a5cd6e3259223b16a2314c5676b88ae1c1f8579a8f';
        const addresses = [addressAssociatedWithUTxOWithNoAssetName];
        const res = await provider.utxoByAddresses({ addresses: toCardanoAddresses(addresses) });
        const txOut: Cardano.TxOut = res[0][1];

        expect(txOut.value.assets!.get(Cardano.AssetId(assetWithNoNameId))).toBeDefined();
        expect(txOut.value.assets!.size).toEqual(2);
      });
    });
  });
});
