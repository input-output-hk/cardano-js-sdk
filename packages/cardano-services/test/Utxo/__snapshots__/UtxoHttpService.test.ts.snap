// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UtxoHttpService healthy state /utxo-by-addresses return UTxOs for a single address 1`] = `
Array [
  Array [
    Object {
      "address": "addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7",
      "index": 0,
      "txId": "19251f57476d7af2777252270413c01383d9503110a68b4fde1a239c119c4f5d",
    },
    Object {
      "address": "addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7",
      "value": Object {
        "coins": 1497246388n,
      },
    },
  ],
]
`;

exports[`UtxoHttpService healthy state /utxo-by-addresses return UTxOs for multiple addresses 1`] = `
Array [
  Array [
    Object {
      "address": "addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7",
      "index": 0,
      "txId": "19251f57476d7af2777252270413c01383d9503110a68b4fde1a239c119c4f5d",
    },
    Object {
      "address": "addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7",
      "value": Object {
        "coins": 1497246388n,
      },
    },
  ],
  Array [
    Object {
      "address": "addr_test1qp620qa3rqzd5fxj3hy4dughv7xx2dt9gu9de70jf8hagdcvmqt35f2psxv7ajj5jnh4ajlc752rert8f9msffxdl45qyjefw8",
      "index": 2,
      "txId": "a08e293256180196de69ff025321be2f7ce5911ce4e3a96406b7c62de0e19bb2",
    },
    Object {
      "address": "addr_test1qp620qa3rqzd5fxj3hy4dughv7xx2dt9gu9de70jf8hagdcvmqt35f2psxv7ajj5jnh4ajlc752rert8f9msffxdl45qyjefw8",
      "value": Object {
        "assets": Map {
          "126b8676446c84a5cd6e3259223b16a2314c5676b88ae1c1f8579a8f" => 70n,
          "22c3b86a5b88a78b5de52f4aed2831d1483b3b7681f1ee2569538130cb6de817be83e93a5771e6c98ee305e1012bae70d9eb9dff70e2d71e4b611779" => 780407n,
        },
        "coins": 2000000n,
      },
    },
  ],
  Array [
    Object {
      "address": "addr_test1qryz24mkq35j8s67fdrm44pe8na7n3tqkmyzy3sgnjq3d7szlx56h6fkjl8y3p73zpyce04eku9w943rcr6rgznp8cwq2axy9q",
      "index": 3,
      "txId": "a08e293256180196de69ff025321be2f7ce5911ce4e3a96406b7c62de0e19bb2",
    },
    Object {
      "address": "addr_test1qryz24mkq35j8s67fdrm44pe8na7n3tqkmyzy3sgnjq3d7szlx56h6fkjl8y3p73zpyce04eku9w943rcr6rgznp8cwq2axy9q",
      "value": Object {
        "assets": Map {
          "22c3b86a5b88a78b5de52f4aed2831d1483b3b7681f1ee2569538130cb6de817be83e93a5771e6c98ee305e1012bae70d9eb9dff70e2d71e4b611779" => 11598880n,
        },
        "coins": 4854535n,
      },
    },
  ],
]
`;

exports[`UtxoHttpService healthy state /utxo-by-addresses returns UTxOs containing multiple assets 1`] = `
Array [
  Array [
    Object {
      "address": "addr_test1qp620qa3rqzd5fxj3hy4dughv7xx2dt9gu9de70jf8hagdcvmqt35f2psxv7ajj5jnh4ajlc752rert8f9msffxdl45qyjefw8",
      "index": 2,
      "txId": "a08e293256180196de69ff025321be2f7ce5911ce4e3a96406b7c62de0e19bb2",
    },
    Object {
      "address": "addr_test1qp620qa3rqzd5fxj3hy4dughv7xx2dt9gu9de70jf8hagdcvmqt35f2psxv7ajj5jnh4ajlc752rert8f9msffxdl45qyjefw8",
      "value": Object {
        "assets": Map {
          "126b8676446c84a5cd6e3259223b16a2314c5676b88ae1c1f8579a8f" => 70n,
          "22c3b86a5b88a78b5de52f4aed2831d1483b3b7681f1ee2569538130cb6de817be83e93a5771e6c98ee305e1012bae70d9eb9dff70e2d71e4b611779" => 780407n,
        },
        "coins": 2000000n,
      },
    },
  ],
  Array [
    Object {
      "address": "addr_test1qryz24mkq35j8s67fdrm44pe8na7n3tqkmyzy3sgnjq3d7szlx56h6fkjl8y3p73zpyce04eku9w943rcr6rgznp8cwq2axy9q",
      "index": 3,
      "txId": "a08e293256180196de69ff025321be2f7ce5911ce4e3a96406b7c62de0e19bb2",
    },
    Object {
      "address": "addr_test1qryz24mkq35j8s67fdrm44pe8na7n3tqkmyzy3sgnjq3d7szlx56h6fkjl8y3p73zpyce04eku9w943rcr6rgznp8cwq2axy9q",
      "value": Object {
        "assets": Map {
          "22c3b86a5b88a78b5de52f4aed2831d1483b3b7681f1ee2569538130cb6de817be83e93a5771e6c98ee305e1012bae70d9eb9dff70e2d71e4b611779" => 11598880n,
        },
        "coins": 4854535n,
      },
    },
  ],
]
`;
